tools/lsp/language.rs

```diff
@@ -576,33 +596,185 @@ pub fn show_preview_command(
+    }
+
+    Ok(serde_json::to_value(()).expect("Failed to serialize ()!"))
+}
+
 pub(crate) async fn reload_document_impl(
     ctx: Option<&Rc<Context>>,
-    mut content: String,
+    content: String,
     url: lsp_types::Url,
     version: Option<i32>,
     document_cache: &mut common::DocumentCache,
 ) -> HashMap<Url, Vec<lsp_types::Diagnostic>> {
+    enum FileAction {
+        ProcessContent(String),
+        IgnoreFile,
+        InvalidateFile,
+    }
+
     let Some(path) = common::uri_to_file(&url) else { return Default::default() };
     // Normalize the URL
     let Ok(url) = Url::from_file_path(path.clone()) else { return Default::default() };
-    if path.extension().map_or(false, |e| e == "rs") {
-        content = match i_slint_compiler::lexer::extract_rust_macro(content) {
-            Some(content) => content,
+
+    let action = if path.extension().map_or(false, |e| e == "rs") {
+        match i_slint_compiler::lexer::extract_rust_macro(content) {
+            Some(content) => FileAction::ProcessContent(content),
             // A rust file without a rust macro, just ignore it
-            None => return [(url, vec![])].into_iter().collect(),
-        };
-    }
+            None => {
+                if document_cache.get_document(&url).is_some() {
+                    // This had contents before: Continue so we can invalidate it!
+                    FileAction::InvalidateFile
+                } else {
+                    FileAction::IgnoreFile
+                }
+            }
+        }
+    } else {
+        FileAction::ProcessContent(content)
+    };
 
-    if let Some(ctx) = ctx {
-        ctx.server_notifier.send_message_to_preview(common::LspToPreviewMessage::SetContents {
-            url: common::VersionedUrl::new(url.clone(), version),
-            contents: content.clone(),
-        });
-    }
-    let dependencies = document_cache.invalidate_url(&url);
     let mut diag = BuildDiagnostics::default();
-    let _ = document_cache.load_url(&url, version, content, &mut diag).await; // ignore url conversion errors
+
+    let dependencies = match action {
+        FileAction::ProcessContent(content) => {
+            if let Some(ctx) = ctx {
+                ctx.server_notifier.send_message_to_preview(
+                    common::LspToPreviewMessage::SetContents {
+                        url: common::VersionedUrl::new(url.clone(), version),
+                        contents: content.clone(),
+                    },
+                );
+            }
+            let dependencies = document_cache.invalidate_url(&url);
+            let _ = document_cache.load_url(&url, version, content, &mut diag).await;
+            dependencies
+        }
+        FileAction::IgnoreFile => return Default::default(),
+        FileAction::InvalidateFile => {
+            if let Some(ctx) = ctx {
+                ctx.server_notifier.send_message_to_preview(
+                    common::LspToPreviewMessage::FileLost { url: url.clone() },
```

Comment:
FileLost?  How can we lose a file? 😉

Reply:
It's not deleted, it just suddenly became unintersting to the LSP. I'll think of a better name.